File: lib/ts/src/internal/types/TMockStub.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { TTypeDeclaration } from './TTypeDeclaration.js';
import { ValidationFailure } from '../validation/ValidationFailure.js';
import { TType } from './TType.js';
import { validateMockStub } from '../validation/ValidateMockStub.js';
import { GenerateContext } from '../generation/GenerateContext.js';
import { generateRandomMockStub } from '../generation/GenerateRandomMockStub.js';
import { ValidateContext } from '../validation/ValidateContext.js';

export const mockStubName: string = '_ext.Stub_';

export class TMockStub extends TType {
    types: { [key: string]: TType };

    constructor(types: { [key: string]: TType }) {
        super();
        this.types = types;
    }

    getTypeParameterCount(): number {
        return 0;
    }

    validate(givenObj: any, typeParameters: TTypeDeclaration[], ctx: ValidateContext): ValidationFailure[] {
        return validateMockStub(givenObj, this.types, ctx);
    }

    generateRandomValue(
        blueprintValue: any,
        useBlueprintValue: boolean,
        typeParameters: TTypeDeclaration[],
        ctx: GenerateContext,
    ): any {
        return generateRandomMockStub(this.types, ctx);
    }

    getName(): string {
        return mockStubName;
    }
}

File: lib/go/internal/types/TMockStub.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package types

const mockStubName = "_ext.Stub_"

// TMockStub represents an external mock stub type.
type TMockStub struct {
	Types map[string]TType
}

// NewTMockStub constructs a TMockStub instance.
func NewTMockStub(types map[string]TType) *TMockStub {
	return &TMockStub{Types: types}
}

// GetTypeParameterCount implements TType.
func (t *TMockStub) GetTypeParameterCount() int {
	return 0
}

// Validate implements TType.
func (t *TMockStub) Validate(value any, typeParameters []*TTypeDeclaration, ctx *ValidateContext) []*ValidationFailure {
	return ValidateMockStub(value, t.Types, ctx)
}

// GenerateRandomValue implements TType.
func (t *TMockStub) GenerateRandomValue(blueprintValue any, useBlueprintValue bool, typeParameters []*TTypeDeclaration, ctx *GenerateContext) any {
	return GenerateRandomMockStub(t.Types, ctx)
}

// GetName implements TType.
func (t *TMockStub) GetName(ctx *ValidateContext) string {
	return mockStubName
}
